symengine/fields.cpp

```diff
@@ -0,0 +1,272 @@
+                coeff *= inv;
+                ++iter;
+            } else {
+                i--;
+            }
+            mp_fdiv_r(dict_out[it], coeff, modulo_);
+            if (it <= 0)
+                break;
+        }
+        map_uint_mpz dict_rem, dict_quo;
+        for (auto it : dict_out) {
+            if (it.first < deg_divisor)
+                dict_rem.insert(dict_rem.end(),{it.first, it.second});
+            else
+                dict_quo.insert(dict_quo.end(),{it.first - deg_divisor, it.second});
+        }
+        *quo = GaloisFieldDict(dict_quo, modulo_);
+        *rem = GaloisFieldDict(dict_rem, modulo_);
+    }
+}
+
+GaloisFieldDict GaloisFieldDict::gf_lshift(const integer_class n) const
+{
+    map_uint_mpz dict_out;
+    unsigned n_val = mp_get_si(n);
+    for (auto iter : dict_) {
+        dict_out.insert(dict_out.end(), {iter.first + n_val, iter.second});
+    }
+    return GaloisFieldDict(dict_out, modulo_);
+}
+
+void GaloisFieldDict::gf_rshift(const integer_class n,
+                            const Ptr<GaloisFieldDict> &quo,
+                            const Ptr<GaloisFieldDict> &rem) const
+{
+    map_uint_mpz dict_quo, dict_rem;
+    unsigned n_val = mp_get_si(n);
+    for (auto iter : dict_) {
+        if (iter.first >= n_val)
+            dict_quo.insert(dict_quo.end(), {iter.first - n_val, iter.second});
+        else
+            dict_rem.insert(dict_rem.end(), iter);
+    }
+    *quo = GaloisFieldDict(dict_quo, modulo_);
+    *rem = GaloisFieldDict(dict_rem, modulo_);
+}
+
+GaloisFieldDict GaloisFieldDict::gf_sqr() const
+{
+    return (*this * *this);
+}
+
+GaloisFieldDict GaloisFieldDict::gf_pow(const integer_class n) const
+{
+    if (n == 0) {
+        return GaloisFieldDict({integer_class(1)}, modulo_);
+    }
+    if (n == 1)
+        return static_cast<GaloisFieldDict>(*this);
+    if (n == 2)
+        return gf_sqr();
+    long num = mp_get_si(n);
+    GaloisFieldDict to_sq = static_cast<GaloisFieldDict>(*this);
+    GaloisFieldDict to_ret = GaloisFieldDict({integer_class(1)}, modulo_);
+    while (1) {
+        if (num & 1) {
+            to_ret *= to_sq;
+        }
+        num >>= 1;
+        if (num == 0)
+            return to_ret;
+        to_sq = to_sq.gf_sqr();
+    }
+}
+
+void GaloisFieldDict::gf_monic(integer_class &res,
+                           const Ptr<GaloisFieldDict> &monic) const
+{
+    if (dict_.empty()) {
+        res = integer_class(0);
+        *monic = static_cast<GaloisFieldDict>(*this);
+    } else {
+        res = dict_.rbegin()->second;
+        if (res == integer_class(1)) 
+            *monic = static_cast<GaloisFieldDict>(*this);
+        else {
+            monic->dict_.clear();
+            monic->modulo_ = modulo_;
+            integer_class inv, temp;
+            mp_invert(inv, res, modulo_);
+            for (auto &iter : dict_) {
+                mp_fdiv_r(temp, inv * iter.second, modulo_);
+                if (temp != integer_class(0))
```

Comment:
Since `gcd(inv, modulo_) == 1` and `iter.second` != 0`, `temp != 0`. This check is unnecessary.